google-cloud-clients/google-cloud-contrib/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStoragePath.java

```diff
@@ -83,8 +86,39 @@ boolean seemsLikeADirectory() {
     return path.seemsLikeADirectory();
   }
 
-  boolean seemsLikeADirectoryAndUsePseudoDirectories() {
-    return path.seemsLikeADirectory() && fileSystem.config().usePseudoDirectories();
+  boolean seemsLikeADirectoryAndUsePseudoDirectories(Storage storage) {
+    if (!fileSystem.config().usePseudoDirectories()) {
```

Comment:
If I turn off the pseudo directories option, then provide a path like `foo/bar/`, the will now return `false` when it should probably return `true`. We should probably swap the conditional order back.

Reply:
The spec is: if you enable pseudo directories, then we give the user the illusion that GCS has folders. So if for example there's a file on GCS with name "foo/bar/hello.txt" then we pretend that folders "foo" and "foo/bar" exist, even though they're really just prefixes of a file name. If we disable them then we tell the user exactly what GCS tells us. Currently, this means that if you ask for "foo" you'll get a "file not found" error (since there's only one file and it's called "foo/bar/hello.txt").

So with pseudo directories off, this function is correct in returning false always - the function is meant to only return true if (as per the name) the given path looks like a directory and pseudo-directories are enabled. We should not swap the conditional order.

That said, I did move conditionals about because I realized it was a bad idea to have the extra list call at the start of every stat call. Now we only do this when we can't find the file.

Reply:
Ok, SGTM then.